README.md

```diff
@@ -1,2 +1,182 @@
+
+### Data structures
+```go
+type Job struct {
+	Cmd    *exec.Cmd
+	Pid    int
+
+	status JobStatus	// Status of this Job
+	*sync.Mutex			// to protect Status from simultaneous reads and writes
+	outStream Streamer	// used for creating streaming channel via Output()
+	doneC     chan struct{} // channel for user to block until job exits
+}
+
+type JobStatus {
+	Status string  // string representation of current status
+	Exited bool    // true if job has exited
+	ExitCode int   // exit code of the exited job
+	// extend as needed
+}
+
+type Streamer interface {
+	io.WriteCloser	// implements Writer and Closer
+	NewReader() (io.ReadCloser, error)
+}
+
+// streamer structure embeds os.File for implementing
+// simultaneous output streaming for multiple clients.
+type streamer struct {
+	*os.File // embeds os.File
+}
+
+```
+  Note that `streamer` structure embeds `os.File` and has many benefits:
+  1. Read() and Write() methods for free
+  2. Synchronization by kernel for free
+  3. Memory-error safe as data is stored in disk
+  4. Will be asigned to exec.Cmd.Stdout and exec.Cmd.Stderr fields and both outputs will go into a single file in chronological order.
+
+Say, if a client is requesting for a job that builds a project, logs can be very long and may take up a lof of space. The above approach is very handy as it uses disk space rather than in-memory buffers. And implementation will be simple as many things come for free.
+
+  
+  To keep it simple, this library will do followings which **might not** be suitable for production:
+1. Execute all provided commands only through `/bin/sh`.
+2. Server (as library user) will track all jobs created in-memory via a map data structure. To make it more practicle, other options can be used (database, file system, infra provided by kernel, etc.). In my opinion, for *Unix* and *Linux*, `/proc/` directory has been one of the most consistent performers for job/task/process monitoring.
+
+### Using library
+Following are the functions this library will provide
+```go
+// NewJob creates a job with specified command but does not start it
+func NewJob(command string) *Job
+// Start starts the job. It will call Start() and Wait() on underlying
+// exec.Cmd. Any error will be reported to job's status. (see data structures)
+func (j* Job) Start() {}
+// Status returns status of this job. See struct JobStatus for more details
+// Users can determine whether specified job has completed or not via this function
+func (j *Job) Status() JobStatus {}
+// Stop terminates a running job. Does nothing if job has already stopped.
+func (j *Job) Stop() error {}
+// Output returns a string channel which will be closed by library internally
+// once the job completes. Multiple calls to this function will create separate channels
+// so it is safe to stream output to multiple clients.
+func (j *Job) Output() chan string {}
```

Comment:
What if the output is not UTF-8 text?

Reply:
I was relying on this information:
`To summarize, strings can contain arbitrary bytes, but when constructed from string literals, those bytes are (almost always) UTF-8.`
https://go.dev/blog/strings : Section **UTF-8 and string literals**
As data is read from a file (not string literals), I think the strings would contain non UTF-8 bytes as well. There should not be a problem. Will add a test case for this scenario to confirm.
If it does not work, I will try using something else without complicating the implementation (may be []byte).

Reply:
I was definitely thinking in terms of user expectations. Converting arbitrary bytes into a `string` _will_ compile, but I think calling it a `string` makes some implicit assurances about what sort of data it will contain.

Reply:
changed `chan string` to `chan []byte` and a very minor change here and there actually made it work as before. Committed to `client_server` branch